cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,304 @@
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// Interval of synchronizing service status from apiserver.
+	serviceSyncPeriod = 30 * time.Second
+	// Frequency at which the firewall controller runs.
+	firewallReconcileRequency = 5 * time.Minute
+)
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	mu       *sync.RWMutex // protects firewall.
+	firewall *godo.Firewall
+}
+
+// firewallManagerOp manages the interaction with the DO Firewalls API.
+type firewallManagerOp struct {
+	client             *godo.Client
+	fwCache            firewallCache
+	workerFirewallName string
+	tags               []string
+}
+
+// FirewallManager retrieves and stores firewall representations.
+type FirewallManager interface {
+	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
+	// if it exists and nil if it does not exist.
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, inboundRules []godo.InboundRule) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient    clientset.Interface
+	client        *godo.Client
+	tags          []string
+	serviceLister corelisters.ServiceLister
+	fwManager     FirewallManager
+	fwCache       firewallCache
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	tags []string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient: kubeClient,
+		client:     client,
+		tags:       tags,
+		fwManager:  fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
```

Comment:
This context will be cancelled before the second iteration of the `Until`, since `serviceSyncPeriod < firewallReconcileFrequency`. Should it be defined inside the closure?

Reply:
Like this @adamwg?
```go
// Run starts the firewall controller loop.
func (fc *FirewallController) Run(stopCh <-chan struct{}) {
	wait.Until(func() {
		ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
		defer cancel()
```